rust/benches/caching.rs

```diff
@@ -0,0 +1,341 @@
+                        .unwrap();
+                }
+                Cache::CachedNodeManager(node_manager) => {
+                    let _unused = node_manager.get_read_access(i).unwrap();
+                }
+            }
+        }
+    }
+
+    /// Returns the capacity of the cache.
+    fn capacity(&self) -> u64 {
+        match self {
+            Cache::QuickCache(cache) => cache.capacity(),
+            Cache::CachedNodeManager(node_manager) => node_manager.capacity(),
+            Cache::LockCache(lock_cache) => lock_cache.capacity(),
+        }
+    }
+
+    /// Executes a read operation on the cache for the given id.
+    fn read(&self, id: u64) {
+        match self {
+            Cache::QuickCache(cache) => match cache.get_value_or_guard(&id, None) {
+                quick_cache::sync::GuardResult::Value(_) => {}
+                quick_cache::sync::GuardResult::Guard(_)
+                | quick_cache::sync::GuardResult::Timeout => {
+                    panic!("Cache miss on QuickCache for id {id}");
+                }
+            },
+            Cache::CachedNodeManager(node_manager) => {
+                let _node = node_manager.get_read_access(id).unwrap();
+            }
+            Cache::LockCache(lock_cache) => {
+                let _node = lock_cache
+                    .get_read_access_or_insert(id, || Ok(42i64))
+                    .unwrap();
+            }
+        }
+    }
+
+    fn add(&self, id: u64) {
+        match self {
+            Cache::QuickCache(cache) => match cache.get_value_or_guard(&id, None) {
+                quick_cache::sync::GuardResult::Guard(guard) => {
+                    guard.insert(42).unwrap();
+                }
+                quick_cache::sync::GuardResult::Timeout
+                | quick_cache::sync::GuardResult::Value(_) => {
+                    panic!("Cache hit on QuickCache for id {id}");
+                }
+            },
+            Cache::CachedNodeManager(node_manager) => {
+                let _node = node_manager.add(42).unwrap();
+            }
+            Cache::LockCache(lock_cache) => {
+                // No real way to enforce an insert
+                let _unused = lock_cache.get_read_access_or_insert(id, || Ok(42)).unwrap();
+            }
+        }
+    }
+}
+
+/// Benchmark caches read performance.
+/// It varies:
+/// - Cache size (influence contention)
+/// - Whether the accessed ids are in cache or not (cache hit/miss)
+/// - Number of threads (influence contention)
+fn read_benchmark(c: &mut criterion::Criterion) {
+    fastrand::seed(123);
+
+    let cache_sizes = [100_000, 1_000_000];
+    for cache_size in cache_sizes {
+        let mut bench_group = c.benchmark_group(format!("caching/read/{cache_size}capacity"));
+        for num_threads in pow_2_threads() {
+            for cache_type in [
+                CacheType::QuickCache,
+                CacheType::LockCache,
+                CacheType::CachedNodeManager,
+            ] {
```

Comment:
Consider adding a `fn variants() -> impl IntoIterator<Item = CacheType>` to `CacheType` and using it here.